sys/net/application_layer/nanocoap/cache.c

```diff
@@ -0,0 +1,327 @@
+       option-specific processing as defined by the option.)
+    */
+    else if (resp->hdr->code == COAP_CODE_VALID) {
+        if (ce) {
+            /* refresh max_age() */
+            uint32_t max_age = 60;
+            coap_opt_get_uint(resp, COAP_OPT_MAX_AGE, &max_age);
+            ce->max_age = ztimer_now(ZTIMER_SEC) + max_age;
+        }
+        /* TODO: handle the copying of the new options (if changed) */
+    }
+    /* This response is not cacheable. However, a cache MUST mark any
+       stored response for the changed resource as not fresh.
+    */
+    else if (resp->hdr->code == COAP_CODE_CHANGED) {
+        if (ce) {
+            /* set max_age to now(), so that the cache is considered
+             * stale immdiately */
+            ce->max_age = ztimer_now(ZTIMER_SEC);
+        }
+    }
+    /* This response is cacheable: Caches can use the Max-Age Option
+       to determine freshness and (if present) the
+       ETag Option for validation.
+    */
+    else if (resp->hdr->code == COAP_CODE_CONTENT) {
+        uint32_t now = ztimer_now(ZTIMER_SEC);
+        /* cache entry is stale */
+        if (ce && (ce->max_age < now)) {
+            nanocoap_cache_del(ce);
+        }
+        if (NULL == nanocoap_cache_add_by_key(cache_key, request_method,
+                                              resp, resp_len)) {
+            /* no space left in the cache? */
+            return -1;
+        }
+        /* TODO: ETAG handling */
+    }
+
+    return 0;
+}
+
+
+static nanocoap_cache_entry_t *_nanocoap_cache_pop(void)
+{
+    clist_node_t *node;
+
+    node = clist_lpop(&_empty_list_head);
+    if (node != NULL) {
+        return container_of(node, nanocoap_cache_entry_t, node);
+    }
+    else {
+        return NULL;
+    }
+}
+
+nanocoap_cache_entry_t *nanocoap_cache_add_by_key(const uint8_t *cache_key,
+                                                  unsigned request_method,
+                                                  const coap_pkt_t *resp,
+                                                  size_t resp_len)
+{
+    nanocoap_cache_entry_t *ce;
+    ce = nanocoap_cache_key_lookup(cache_key);
+
+    /* found an already existing cache entry */
+    if (ce) {
+        return ce;
+    }
+
+    /* did not find .. get an empty cache container */
+    ce = _nanocoap_cache_pop();
+
+    /* no space left */
+    if (!ce) {
+        /* could not remove any entry */
+        if (_replacement_strategy()) {
+            return NULL;
+        }
+        /* could remove an entry */
+        ce = _nanocoap_cache_pop();
+        if (!ce) {
+            /* still no free space ? stop trying now */
+            return NULL;
+        }
+    }
+
+    memcpy(ce->cache_key, cache_key, CONFIG_NANOCOAP_CACHE_KEY_LENGTH);
+    memcpy(&ce->response_pkt, resp, sizeof(coap_pkt_t));
+    memcpy(&ce->response_buf, resp->hdr, resp_len);
+    ce->response_pkt.hdr = (coap_hdr_t *) ce->response_buf;
```

Comment:
While it isn't used here in this, the payload pointer should also be updated, as option parsing relies on it:

```suggestion
    ce->response_pkt.hdr = (coap_hdr_t *) ce->response_buf;
    ce->response_pkt.payload = ce->response_buf + (resp->payload - ((uint8_t *)resp->hdr));
```